### PR TITLE
New server error codes

### PR DESCRIPTION
**Pull request by harminius**
CU: https://app.clickup.com/t/861mzqxud

Apply new custom errors structure to existing errors.

### CODE REVIEW

File: server/mergin/sync/errors.py [hunk 1]
```diff
@@ -19,3 +19,18 @@ def to_dict(self) -> Dict:
         data["invalid_usernames"] = self.invalid_usernames
         data["invalid_ids"] = self.invalid_ids
         return data
+
+
+class StorageLimitHit(ResponseError):
+    code = "StorageLimitHit"
+    detail = "You have reached a data limit"
+
+    def __init__(self, current_usage: str, storage_limit: str):
```
Comment (varmar05) [inline server/mergin/sync/errors.py:28]: storage are of int type

File: server/mergin/sync/public_api_controller.py [hunk 2]
```diff
@@ -821,8 +824,9 @@ def project_push(namespace, project_name):
     if not ws:
         abort(404)
 
-    if ws.disk_usage() + additional_disk_usage > ws.storage:
-        abort(400, "You have reached a data limit")
+    requested_storage = ws.disk_usage() + additional_disk_usage
+    if requested_storage > ws.storage:
+        abort(make_response(jsonify(StorageLimitHit(requested_storage, ws.storage).to_dict()), 422))
```
Comment (varmar05) [inline server/mergin/sync/public_api_controller.py:829]: why not return jsonify(StorageLimitHit(current_disk_usage, ws.storage).to_dict()) ?

I think we should return current usage, not requested storage

**Comment by varmar05** [comment]
One more thing @harminius: let's add missing check for storage in clone endpoint (destination workspace + source project).

File: server/mergin/sync/public_api_controller.py [hunk 3]
```diff
@@ -1126,6 +1135,17 @@ def clone_project(namespace, project_name):  # noqa: E501
             msg = "Project with the same name already exists"
         abort(409, msg)
 
+    # Check storage limit
+    latest_version = cloned_project.get_latest_version()
+    current_usage = ws.disk_usage()
+    requested_storage = current_usage + latest_version.project_size
```
[the remark this answers is not in the document]
Comment (harminius) [reply server/mergin/sync/public_api_controller.py:1141]: i thought project size is sum of size of all versions, and when you clone a project you get only latest version without history - so your storage will be increased by the size of the latest version,

am i wrong?

[the remark this answers is not in the document]
Comment (varmar05) [reply server/mergin/sync/public_api_controller.py:1141]: nope, project disk usage is only size of all current files (=latest version), full project size (space occupied on disk) would be as you said roughly sum over all versions (it is bit more complicated than that) but that would count to history_storage not workspace.storage

**Review by varmar05 — APPROVED** [review]

**Pull request merged**